workflow/worker.go

```diff
@@ -0,0 +1,174 @@
+	return func(opts *workerOptions) error {
+		opts.daprClient = input
+		return nil
+	}
+}
+
+func NewWorker(opts ...workerOption) (*WorkflowWorker, error) {
+	options := new(workerOptions)
+	for _, configure := range opts {
+		if err := configure(options); err != nil {
+			return nil, errors.New("failed to load options")
+		}
+	}
+	var daprClient dapr.Client
+	var err error
+	if options.daprClient == nil {
+		daprClient, err = dapr.NewClient()
+	} else {
+		daprClient = options.daprClient
+	}
+	if err != nil {
+		return nil, err
+	}
+	grpcConn := daprClient.GrpcClientConn()
+
+	return &WorkflowWorker{
+		tasks:  task.NewTaskRegistry(),
+		client: durabletaskclient.NewTaskHubGrpcClient(grpcConn, backend.DefaultLogger()),
+		quit:   make(chan bool),
+		close:  daprClient.Close,
+	}, nil
+}
+
+// getFunctionName returns the function name as a string
+func getFunctionName(f interface{}) (string, error) {
+	if f == nil {
+		return "", errors.New("nil function name")
+	}
+
+	callSplit := strings.Split(runtime.FuncForPC(reflect.ValueOf(f).Pointer()).Name(), ".")
+
+	funcName := callSplit[len(callSplit)-1]
+
+	if funcName == "1" {
+		return "", errors.New("anonymous function name")
+	}
+
+	return funcName, nil
+}
+
+func wrapWorkflow(w Workflow) task.Orchestrator {
+	return func(ctx *task.OrchestrationContext) (any, error) {
+		wfCtx := &WorkflowContext{orchestrationContext: ctx}
+		return w(wfCtx)
+	}
+}
+
+func (ww *WorkflowWorker) RegisterWorkflow(w Workflow) error {
+	wrappedOrchestration := wrapWorkflow(w)
+
+	// get the function name for the passed workflow
+	name, err := getFunctionName(w)
+	if err != nil {
+		return fmt.Errorf("failed to get workflow decorator: %v", err)
+	}
+
+	err = ww.tasks.AddOrchestratorN(name, wrappedOrchestration)
+	return err
+}
+
+func wrapActivity(a Activity) task.Activity {
+	return func(ctx task.ActivityContext) (any, error) {
+		aCtx := ActivityContext{ctx: ctx}
+
+		return a(aCtx)
+	}
+}
+
+func (ww *WorkflowWorker) RegisterActivity(a Activity) error {
+	wrappedActivity := wrapActivity(a)
+
+	// get the function name for the passed activity
+	name, err := getFunctionName(a)
+	if err != nil {
+		return fmt.Errorf("failed to get activity decorator: %v", err)
+	}
+
+	err = ww.tasks.AddActivityN(name, wrappedActivity)
+	return err
+}
+
+func (ww *WorkflowWorker) Start() error {
+	// go func start
+	errChan := make(chan error)
+	go func() {
+		defer ww.close()
+		ctx, cancel := context.WithCancel(context.Background())
+		err := ww.client.StartWorkItemListener(ctx, ww.tasks)
+		if err != nil {
+			cancel()
+			errChan <- fmt.Errorf("failed to start work stream: %v", err)
+			return
+		}
+		ww.cancel = cancel
+		log.Println("work item listener started")
+		errChan <- nil
+		<-ww.quit
+		log.Println("work item listener shutdown")
+	}()
+	err := <-errChan
+	return err
+}
+
+func (ww *WorkflowWorker) Shutdown() error {
+	ww.cancel()
+	// send close signal
+	ww.quit <- true
```

Comment:
If I'm reading the code correctly, it seems the only thing that invoking `ww.quit` does is cause the message `"work item listener shutdown` to get logged. Rather than using a channel for that, why not just log it here?